vllm/model_executor/layers/attention.py

```diff
@@ -297,7 +349,53 @@ def _paged_attention(
     # For context len > 8192, use V2 kernel to avoid shared memory shortage.
     use_v1 = input_metadata.max_context_len <= 8192 and (
         max_num_partitions == 1 or num_seqs * num_heads > 512)
-    if use_v1:
+
+    # NOTE: in `_prepare_prompt` and `_prepare_decode` is filled in different
+    # manner (which may needs to be fixed in the future): in the former
+    # `context_lens` is the length of contexts whose kv-cache has been stored
+    # (e.g., with prefix cache), however, in the later `context_lens` is the
+    # length of current attention context (includes the token whose kv-cache
+    # will be filled in this round).
+    #
+    # The kernel `flash_attn_with_kvcache` expects `cache_seqlens` to be the
+    # length of the context whose kv-cache has been stored, i.e., the value of
+    # `context_lens - seq_len` for decoding.
+    #
+    # The `context_attention_fwd` kernel expects the same samatics as the
+    # `flash_attn_with_kvcache` kernel.
+    #
+    # In the contrast, both `paged_attention_v1` and `paged_attention_v2` expects
+    # the `context_lens` to be the length of the current attention context.
+
+    if input_metadata.use_flash_attn:
+        # see also: https://github.com/Dao-AILab/flash-attention/commit/54e80a3829c6d2337570d01e78ebd9529c02d342
+        output = flash_attn_with_kvcache(
```

Comment:
Thanks for sharing the implementation. Is this faster than context_attention_fwd?

Reply:
I haven't benchmark these two kernels, but I have added a end-to-end speculative sampling benchmark in the last part of #3010's pull request description.